File: crawler/crawl_legislator_info.py
from os import path, environ
from sys import exit

import requests

FILE_DIR = path.dirname(path.abspath(__file__))
OUTPUT_DIR = environ.get('OUTPUT_DIR', f'{FILE_DIR}/../data')
URL = 'http://data.ly.gov.tw/odw/openDatasetJson.action'


def send_request(payload):
    print(f'[INFO] Sending request to {URL}')
    try:
        response = requests.get(URL, params=payload, timeout=10)
    except requests.exceptions.ConnectTimeout:
        exit('[ERROR] Request Timeout')

    assert response.status_code == 200, f'[ERROR] Request Error, response code: {response.status_code}'
    return response.text


def store_info(file_name, response_body):
    OUTPUT_PATH = f'{OUTPUT_DIR}/{file_name}'
    with open(OUTPUT_PATH, 'w+') as fp:
        fp.write(response_body)
        print(f'[INFO] Response_body is written to "{OUTPUT_PATH}"')


def crawl_legislator_info_pages(id, payload_base={}, page_count=1):
    for page_num in range(1, page_count + 1):
        print(f'[INFO] Start crawling "{id}" page{page_num} ')
        payload = {**payload_base, 'page': page_num}
        file_name = f'{id}_page{page_num}.json'
        response_body = send_request(payload)
        store_info(file_name, response_body)


if __name__ == "__main__":
    crawl_legislator_info_pages('history_legislator_info', payload_base={'id': 16, 'selectTerm': 'all'}, page_count=2)
    crawl_legislator_info_pages('current_legislator_info', payload_base={'id': 9, 'selectTerm': 'all'}, page_count=1)
    crawl_legislator_info_pages('meeting_info', payload_base={'id': 42, 'selectTerm': 'all'}, page_count=6)

File: crawler/crawl_legislator_info.go
/*
This is the go version of getting the opendata response
*/

package main

import (
	"io/ioutil"
	"log"
	"net/http"
	"fmt"
	"strconv"
)

type Payload struct {
	id         string
	selectTerm string
}

func crawlPage(url string, filePath string) {
	res, err := http.Get(url)
	if err != nil {
		log.Fatal(err)
	}
	defer res.Body.Close()

	responseBody, err := ioutil.ReadAll(res.Body)
	if err != nil {
		log.Fatal(err)
	}

	err = ioutil.WriteFile(filePath, responseBody, 0644)
	if err != nil {
		log.Fatal(err)
	}

	log.Print(fmt.Sprintf("[INFO] Finish crawling %s to %s", url, filePath))
}

func getFullUrl(baseUrl string, payload Payload) string{
	req, err := http.NewRequest("GET", baseUrl, nil)
	if err != nil {
		log.Fatal(err)
	}

	q := req.URL.Query()
	q.Add("id", payload.id)
	q.Add("selectTerm", payload.selectTerm)
	req.URL.RawQuery = q.Encode()

	return req.URL.String()
}

func crawlLegislatorInfoPages(id string, payload Payload, pageNum int) {
	url := "http://data.ly.gov.tw/odw/openDatasetJson.action"
	for page := 1; page <= pageNum; page++ {
		
		fullUrl := getFullUrl(url, payload)
		filePath := fmt.Sprintf("../data/%s_page%s.json", id, strconv.Itoa(page))
		
		crawlPage(fullUrl, filePath)
	}
}

func main() {
	crawlLegislatorInfoPages("history_legislator_info", Payload{id: "16", selectTerm: "all"}, 2)
	crawlLegislatorInfoPages("current_legislator_info", Payload{id: "9", selectTerm: "all"}, 1)
}
